### PR TITLE
Feature/ea server mocker plugin

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/70530

This PR adds an `ea-server-mocker` plugin to `common` tests folder, see the README to understand why there.

This is not a finished tool but it's the base to allow us to manually test the client without having to rely on a **real** EA server; this opens the door to allow end-to-end (acceptance, REST) testing of the client-server communication too.

### CODE REVIEW

File: tests/_support/ea-server-mocker/ea-service-mocker.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	 * Adding a class that uses  options? Add it here.
+	 *
+	 * Keep in mind this is a testing tool and you want to put your nose in
+	 * the database to check for stuff; refrain from using serialized options
+	 * and make your life easier: you do not want to look at nested serialized arrays.
+	 * Your database will survive the burden of the additional options: don't worry.
+	 *
+	 * @var Tribe__Events__Aggregator_Mocker__Option_Provider_Interface[]
+	 */
+	protected static $option_providers
```
Comment (borkweb) [inline tests/_support/ea-server-mocker/ea-service-mocker.php]: How about:

```php
protected static $option_providers = array(
  'Tribe__Events__Aggregator_Mocker__Service_Options',
);
```

It'll more closely match other multi-line standards we have and will save on one level of tabbing. Rinse and repeat this comment for other places this occurs in the plugin.

**Comment by borkweb** [comment]
This is pretty great, @lucatume!

File: tests/_support/ea-server-mocker/ea-service-mocker.php (hunk 1 above)
Comment (lucatume) [inline tests/_support/ea-server-mocker/ea-service-mocker.php]: I won't blame PHPStorm
I won't blame PHPStorm
I won't blame PHPStorm
I won't blame PHPStorm
I won't blame PHPStorm

File: tests/_support/ea-server-mocker/ea-service-mocker.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+			$class_path = str_replace( array( $prefix, '__' ), array( '', DIRECTORY_SEPARATOR ), $class );
+			/** @noinspection PhpIncludeInspection */
+			require dirname( __FILE__ ) . '/src/' . $class_path . '.php';
+		}
+	}
+
+	public function mock() {
+		$this->hook();
+
+		if ( empty( get_option( 'ea_mocker-enable' ) ) ) {
```
Comment (bordoni) [inline tests/_support/ea-server-mocker/ea-service-mocker.php]: Can we avoid this? PHP 5.3 error.

File: tests/_support/ea-server-mocker/ea-service-mocker.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	/**
+	 * Replaces container bindings with mocking implementations.
+	 */
+	public function replace_bindings() {
+		/** @var Tribe__Events__Aggregator_Mocker__Binding_Provider_Interface $provider */
+		foreach ( $this->bindings_providers as $provider ) {
+			$enable_on = $provider::enable_on();
+			$enabled = false;
+			if ( $enable_on === true ) {
```
Comment (bordoni) [inline tests/_support/ea-server-mocker/ea-service-mocker.php]: Yoda.

File: tests/_support/ea-server-mocker/ea-service-mocker.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		}
+
+		add_action( 'tribe_events_bound_implementations', array( $this, 'replace_bindings' ) );
+	}
+
+	protected function hook() {
+		add_action( 'init', array( new Tribe__Events__Aggregator_Mocker__Options_Page(), 'hook' ) );
+		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+
+		if ( empty( get_option( 'ea_mocker-enable' ) ) ) {
```
Comment (bordoni) [inline tests/_support/ea-server-mocker/ea-service-mocker.php]: Avoid this.

File: tests/_support/ea-server-mocker/src/Options_Page.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+class Tribe__Events__Aggregator_Mocker__Options_Page {
+
+	public function hook() {
+		add_action( 'admin_init', array( $this, 'register_settings' ) );
+		add_action( 'admin_menu', array( $this, 'register_menu' ) );
+	}
+
+	public function register_menu() {
+		add_menu_page( 'Event Aggregator Server Mocker', 'EA Mocker', 'administrator', 'ea-mocker', array( $this, 'render' ) );
```
Comment (bordoni) [inline tests/_support/ea-server-mocker/src/Options_Page.php]: Break it into Multiple lines?

**Comment by lucatume** [comment]
Code style issues fixed

File: tests/_support/ea-server-mocker/src/Service.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		);
+	}
+
+	/**
+	 * Fetch origins from service
+	 *
+	 * @return array
+	 */
+	public function get_origins() {
+		if ( ! empty( get_option( 'ea_mocker-origins-mock_response' ) ) ) {
```
Comment (bordoni) [inline tests/_support/ea-server-mocker/src/Service.php]: Avoid this, will break on PHP 5.3<

File: tests/_support/ea-server-mocker/src/Service.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+	/**
+	 * Fetch import data from service
+	 *
+	 * @param string $import_id ID of the Import Record
+	 *
+	 * @return stdClass|WP_Error
+	 */
+	public function get_import( $import_id ) {
+		if ( ! empty( json_decode( get_option( 'ea_mocker-import-mock_response' ) ) ) ) {
```
Comment (bordoni) [inline tests/_support/ea-server-mocker/src/Service.php]: Same here, PHP 5.2 compat. Also would be cool to do the opposite, return something if it's `empty` and in the end return what we want.

Comment (lucatume) [inline tests/_support/ea-server-mocker/src/Service.php]: What is missing, and I've added, is returning the original method when the mock should not happen...

**Comment by bordoni** [comment]
Merge away.

**Pull request merged**